### PR TITLE
Use Secure vault in ClientAuthenticationHandler

### PR DESCRIPTION
**Pull request by WinmaH**
### Proposed changes in this pull request
This PR will add Secure Vault support in ClientAUthrnticationHandler

#### Documentation
Information on the proposed change can be found at
( https://docs.google.com/document/d/1JAys6iW3L5jEW6k7ghjvYTHsrjYe0iRLxEiAldSeBj8/edit?usp=sharing)

### CODE REVIEW

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 1]
```diff
@@ -26,6 +28,7 @@
     private List<String> intermediateCertCNList = new ArrayList<>();
     private List<String> exemptedContextList = new ArrayList<>();
     private boolean isIntermediateCertValidationEnabled = false;
+    private SecretResolver secretResolver;
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: should this be a class varibale? can't we move that to the method.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 2]
```diff
@@ -140,8 +143,13 @@ public void buildClientAuthenticationHandlerControlData() {
             if ( applications != null ) {
                 while ( applications.hasNext() ) {
                     OMElement resource = applications.next();
+                    setSecretResolver(resource);
                     String appName = resource.getAttributeValue(new QName(Constants.APPLICATION_NAME_ATTR));
                     String hash = resource.getAttributeValue(new QName(Constants.APPLICATION_HASH_ATTR));
+                    if (secretResolver != null && secretResolver.isInitialized()
+                            && secretResolver.isTokenProtected("ClientApp.Authentication.Hash")) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: use a constant. ClientApp.Authentication.Hash

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java (hunk 1 above)
Comment (WinmaH) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: > should this be a class varibale? can't we move that to the method.

moved inside the class variable

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java (hunk 2 above)
Comment (WinmaH) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: > use a constant. ClientApp.Authentication.Hash

fixed

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 3]
```diff
@@ -131,6 +134,7 @@ public void buildResourceAccessControlData() {
      */
     public void buildClientAuthenticationHandlerControlData() {
 
+        SecretResolver secretResolver;
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: why do we need this outside of the while loop?

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 4]
```diff
@@ -140,8 +144,13 @@ public void buildClientAuthenticationHandlerControlData() {
             if ( applications != null ) {
                 while ( applications.hasNext() ) {
                     OMElement resource = applications.next();
+                    secretResolver = SecretResolverFactory.create(resource, true);
```
Comment (WinmaH) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: initialize the secretResolver on the same line

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java (hunk 3 above)
Comment (WinmaH) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: > why do we need this outside of the while loop?
fixed

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 5]
```diff
@@ -140,8 +143,13 @@ public void buildClientAuthenticationHandlerControlData() {
             if ( applications != null ) {
                 while ( applications.hasNext() ) {
                     OMElement resource = applications.next();
+                    SecretResolver secretResolver = SecretResolverFactory.create(resource, true);
                     String appName = resource.getAttributeValue(new QName(Constants.APPLICATION_NAME_ATTR));
                     String hash = resource.getAttributeValue(new QName(Constants.APPLICATION_HASH_ATTR));
+                    if (secretResolver.isInitialized()
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: Let's add appropriate debug logs. When the secret resolver is not initialized and when the token is / is not protected.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 6]
```diff
@@ -26,6 +28,7 @@
     private List<String> intermediateCertCNList = new ArrayList<>();
     private List<String> exemptedContextList = new ArrayList<>();
     private boolean isIntermediateCertValidationEnabled = false;
+    private final String CLIENT_APP_AUTHENTICATION_DASHBOARD_HASH = "ClientApp.Authentication.Dashboard.Hash";
```
[the remark this answers is not in the document]
Comment (WinmaH) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: added the functionality to resolve secrets for any client application

[the remark this answers is not in the document]
Comment (WinmaH) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: > we can't hardcode Dashboard here? It should be configurable

added the functionality to resolve secrets for any client application

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java (hunk 5 above)
Comment (WinmaH) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: > Let's add appropriate debug logs. When the secret resolver is not initialized and when the token is / is not protected.

added debug logs

**Pull request merged**